builder/pipeline.cc

```diff
@@ -886,29 +886,81 @@ class pipeline_builder {
       for (const auto& i : f->inputs()) {
         const auto& input = i.buffer;
 
-        if (!input->producer()) {
-          continue;
-        }
-
-        if (!f->impl() && !f->padding()) {
+        if ((!f->impl() && !f->padding()) || (input->constant())) {
           // Collect all buffers which are inputs to the copy
           // and the outputs of the copy as their dependencies.
           copy_inputs_.insert(input->sym());
           for (const func::output& o : f->outputs()) {
             const buffer_expr_ptr& b = o.buffer;
-            if (output_syms_.count(b->sym())) continue;
+            if (!input->constant() && output_syms_.count(b->sym())) continue;
             copy_deps_[input->sym()].push_back(b->sym());
           }
         }
 
-        allocation_info_[input->sym()]->deps_count++;
+        if (allocation_info_[input->sym()]) {
+          allocation_info_[input->sym()]->deps_count++;
+        }
       }
     }
   }
 
+  void place_constrained_buffers(std::vector<statement_with_range>& results, std::set<var>& candidates,
+      const std::vector<allocation_candidate>& special, std::vector<allocation_candidate>& new_special,
+      symbol_map<var>& uncropped_subs) {
+    for (std::size_t iy = 0; iy < special.size(); iy++) {
+      bool found = false;
+      for (std::size_t ix = 0; ix < results.size(); ix++) {
+        buffer_expr_ptr candidate = special[iy].buffer;
+
+        bool is_ready = std::all_of(copy_deps_[candidate->sym()].begin(), copy_deps_[candidate->sym()].end(),
+            [&](var b) { return results[ix].allocations.count(b) > 0; });
+
+        if (!is_ready) {
+          continue;
+        }
+
+        // The block range must fully cover the allocation range.
+        if (results[ix].start <= special[iy].lifetime_start && special[iy].lifetime_end <= results[ix].end) {
```

Comment:
If constants had an appropriate value of `lifetime_start` and `lifetime_end`, couldn't they be handled by this code without a special case? `produce_allocation` can easily be modified to produce either an `allocate` or a `constant_buffer`.

Reply:
I feel like this is going to add much more code and make it more confusing. I moved common part of these function into a helper, I think it's slightly better now.